src/Features/LanguageServer/Protocol/Features/Diagnostics/DefaultDiagnosticAnalyzerService.cs

```diff
@@ -38,7 +38,16 @@ internal partial class DefaultDiagnosticAnalyzerService : IIncrementalAnalyzerPr
         }
 
         public IIncrementalAnalyzer CreateIncrementalAnalyzer(Workspace workspace)
-            => new DefaultDiagnosticIncrementalAnalyzer(this, workspace);
+        {
+            if (_globalOptions.IsPullDiagnostics(InternalDiagnosticsOptions.NormalDiagnosticMode))
+            {
+                // We rely on LSP to query us for diagnostics when things have changed and poll us for changes that might
+                // have happened to the project or closed files outside of VS.
```

Comment:
Note that `DefaultDiagnosticAnalyzerService` does not support pull model for getting diagnostics: https://github.com/dotnet/roslyn/blob/5b9a9d44eee29f433d6c9ef9740642dd4d08e963/src/Features/LanguageServer/Protocol/Features/Diagnostics/DefaultDiagnosticAnalyzerService.cs#L46-L53

This diagnostic analyzer service is used when no workspace kinds are provided: https://github.com/dotnet/roslyn/blob/5b9a9d44eee29f433d6c9ef9740642dd4d08e963/src/Features/LanguageServer/Protocol/Features/Diagnostics/DefaultDiagnosticAnalyzerService.cs#L23-L24

While our actual `DiagnosticAnalyzerService` is used for Host and Interactive workspace kinds: https://github.com/dotnet/roslyn/blob/5b9a9d44eee29f433d6c9ef9740642dd4d08e963/src/Features/LanguageServer/Protocol/Features/Diagnostics/DiagnosticAnalyzerService_IncrementalAnalyzer.cs#L14-L17

I am wondering whether `DefaultDiagnosticIncrementalAnalyzer` should depend on the pull diagnostics option or not.

Reply:
> If there is a valid scenario for it in VS

I believe it is primarily for non-VS scenarios, I don't know if it is used for any VS scenarios. AdHocWorkspace, MSBuildWorkspace, etc. should use `DefaultDiagnosticAnalyzerService`.

Reply:
Gotcha.  I created https://github.com/dotnet/roslyn/issues/61442 to track, but at least for the immediate moment it shouldn't block LSP pull diagnostics in VS then (though it will block removing the sln crawler).